timecache/first_seen_cache.go

```diff
@@ -0,0 +1,70 @@
+package timecache
+
+import (
+	"container/list"
+	"sync"
+	"time"
+)
+
+// FirstSeenCache is a thread-safe copy of https://github.com/whyrusleeping/timecache.
+type FirstSeenCache struct {
+	q     *list.List
+	m     map[string]time.Time
+	span  time.Duration
+	guard *sync.RWMutex
+}
+
+func newFirstSeenCache(span time.Duration) TimeCache {
+	return &FirstSeenCache{
+		q:     list.New(),
+		m:     make(map[string]time.Time),
+		span:  span,
+		guard: new(sync.RWMutex),
+	}
+}
+
+func (tc FirstSeenCache) Add(s string) {
+	tc.guard.Lock()
+	defer tc.guard.Unlock()
+
+	_, ok := tc.m[s]
+	if ok {
+		panic("putting the same entry twice not supported")
+	}
+
+	tc.sweep()
```

Comment:
This should probably happen in the backround instead of every add, it is inefficient.

We dont have to do it in this pr, but if we dont we should add a TODO and open an issue for continuation.

Reply:
What are your thoughts around where/how to do this? I could do it in this PR, if not complicated.

Or, if you would prefer to take some time and discuss the implementation, I could add a TODO for later.

If you meant in a goroutine with some kind of timed garbage collection, then would make sense to do in a separate PR.

Reply:
Lets add a TODO and do follow up, discussing in issue.

Reply:
and yes, a separate goroutine periodically running a sweep is what i have in mind.